### ISSUE TITLE
cimgui as a static library with glfw and dx11

### ISSUE THREAD

**Issue by CybernetHacker14** [cimgui/cimgui#266]
Hi, this might be a silly question, but I didn't find a proper answer anywhere - can this library be included as a static library and added to a C project directly?

I am trying to make it work with GLFW and DX11, so far I made a fork and made the changes to the `generator.bat` file. I also recursively cloned the `imgui` submodule, generated the `cimgui_impl.h` header with the new function calls for the backends which I want.
I also wrote a premake file to compile the project as a static library, to be included as a dependency for my main application dll.

However, when I use any of the functions in my c file, it throws a linker error - unresolved external symbol, for every function from `cimgui_impl.h` header which I am using.

Not sure, what I am doing wrong, or whether am I compiling `cimgui` in a wrong manner altogether

**Comment by CybernetHacker14** [comment]
I believe someone was trying almost the same thing I am trying to figure out, and posted a question regarding this on StackOverflow - https://stackoverflow.com/questions/71561558/how-can-i-compile-cimgui-statically-and-then-use-it.

It says it can be only used as a shared library. So is it a hard restriction?

**Comment by sonoro1234** [comment]
> unresolved external symbol, for every function from `cimgui_impl.h` header which I am using.

Did you remembered to compile the cpp file declared in cimgui_impl.h?

**Comment by sonoro1234** [comment]
> It says it can be only used as a shared library. So is it a hard restriction?

No that I know

**Comment by CybernetHacker14** [comment]
> > unresolved external symbol, for every function from `cimgui_impl.h` header which I am using.
> 
> Did you remembered to compile the cpp file with funtions declared in cimgui_impl.h?

Yes I did

**Comment by sonoro1234** [comment]
Imgui_impl_glfw.cpp and imgui_impl_dx11.cpp?
That is odd. Should not get unresolved function as they are present in this files
Perhaps if you first try a compilation with this backends for shared library could help to find errors

**Comment by sonoro1234** [comment]
Also did you defined CIMGUI_USE_DX11 and for glfw also?

**Comment by CybernetHacker14** [comment]
Yes, I used the macros - `CIMGUI_USE_DX11` and `CIMGUI_USE_GLFW`, but it was still throwing those linking errors.

It's a complete oddity, but I was able to make things work by referring to some stuff mentioned in this StackOverflow answer - https://stackoverflow.com/questions/76140849/how-can-i-integrate-cimgui-with-glfw-and-vulkan-into-a-zig-project, along with a couple of my own changes as well. I pretty sure that's not the right way things should be working, but I was able to get it up and running.

I did the following set of changes
- In my `main.c` file, where I was implementing the render loop and integrating and adding cimgui, I removed the included `cimgui_impl.h` header. Also, I removed those two above mentioned macros.
- I also removed the `imgui_impl_glfw.h` and `imgui_impl_dx11.h` headers from the cimgui project.
- I copied the struct typedefs and the function declarations present in the `cimgui_impl.h` and pasted it in the end of the `cimgui.h` header. I also added an `ig` prefix at the beginning of the function names like how it's mentioned in one of the points of the StackOverflow link answer.
- In the `cimgui.cpp` file, I am including the `backends/imgui_impl_glfw.cpp` and `backends/imgui_impl_dx11.cpp` files just before the definitions for function declarations mentioned in the previous point. Inside each function definition, I am calling the appropriate exact ImGui function which needs to be executed. So, the beginning of the change patch should look something like this
```cpp
#include "imgui/backends/imgui_impl_glfw.cpp"
#include "imgui/backends/imgui_impl_dx11.cpp"

CIMGUI_API bool igImGui_ImplGlfw_InitForOther(GLFWwindow* window, bool install_callbacks) {
    return ImGui_ImplGlfw_Init(window, install_callbacks, GlfwClientApi_Unknown);
}

CIMGUI_API void igImGui_ImplGlfw_Shutdown(void) {
    ImGui_ImplGlfw_Shutdown();
}

CIMGUI_API void igImGui_ImplGlfw_NewFrame(void) {
    ImGui_ImplGlfw_NewFrame();
}
```

**Comment by sonoro1234** [comment]
Glad it works for you. The simplest thing would be to base your example on any of the provided backend_test examples.

**Comment by CybernetHacker14** [comment]
I based it earlier exactly on the examples provided in the backend_test folder - but it doesn't work. If I have to edit the `cimgui_impl.h`, `cimgui.h` and `cimgui.cpp` headers manually, it defeats the purpose of an automated, generator-based solution.

**Comment by sonoro1234** [comment]
> If I have to edit the `cimgui_impl.h`, `cimgui.h` and `cimgui.cpp` headers manually, it defeats the purpose of an automated, generator-based solution.

You dont need to do that, just a new generator.bat (sh) and new building system outside of cimgui repo https://github.com/cimgui/cimgui/issues/267

**Comment by CybernetHacker14** [comment]
Still trying to wrap my ahead around why it's throwing unresolved external symbol error, even after of following the instructions mentioned in the README to the point.
This is the branch on which I am currently working and has the cimgui integration - https://github.com/CybernetHacker14/Sentinel/tree/dev-c-refactor-0.0.02

I guess the conclusion would be that, it would only work with the combination of examples provided in the `backend_test` folder, and no other combination should work apart from what's exactly provided.

**Comment by sonoro1234** [comment]
So you want glfw, dx11 and static library. 
First I did generation adding `dx11` to generator.bat. 
Then I took example_glfw_opengl as a start: drop all opengl references and add `list(APPEND IMGUI_SOURCES ${BAKENDS_FOLDER}imgui_impl_dx11.cpp)`

If I comment all `main.c` related things to just generate the static libraries compilation succeeds.

The next task would be to modify main.c in order to compile the executable. I should do it by modifying imgui/examples/example_glfw_dx11 but it does not exist.

**Comment by CybernetHacker14** [comment]
Yeah, `imgui/examples/example_glfw_dx11` won't exist, because it's an oddball combination - not exactly done before.
But it works, because I tried it in a [cpp version of my project](https://github.com/CybernetHacker14/Sentinel/tree/cpp-backup)

What I used to do previously, is create a file named [ImGuiBuild.cpp](https://github.com/CybernetHacker14/Sentinel/blob/cpp-backup/Engine/Source/Platform/Windows/ImGui/ImGuiBuild.cpp) in my project, and include the two backend cpp files - `imgui_impl_dx11.cpp` and `imgui_impl_glfw.cpp`. And in all those places where I needed to execute the functionality, I used to include the headers.

I am guessing something similar won't work out for a C project, making compiling as a static library difficult without changing the generator flow.

**Comment by CybernetHacker14** [comment]
> So you want glfw, dx11 and static library. First I did generation adding `dx11` to generator.bat. Then I took example_glfw_opengl as a start: drop all opengl references and add `list(APPEND IMGUI_SOURCES ${BAKENDS_FOLDER}imgui_impl_dx11.cpp)`
> 
> If I comment all `main.c` related things to just generate the static libraries compilation succeeds.
> 
> The next task would be to modify main.c in order to compile the executable. I should do it by modifying imgui/examples/example_glfw_dx11 but it does not exist.
> 
> I also saw [ocornut/imgui#5178](https://github.com/ocornut/imgui/issues/5178) where it seems you had a main.cpp for that combination. I you could provide such file perhaps I could adapt it to work from C.
> 
> Work already done in https://github.com/cimgui/cimgui/tree/glfw_dx11_example

Is this the one you are looking for - https://github.com/CybernetHacker14/Sentinel/blob/6d5f9d31141d33adf883ebadeb77b5645ab82832/Engine/Source/Sentinel/GUI/ImGui/ImGuiLayer.cpp

**Comment by sonoro1234** [comment]
> Is this the one you are looking for

No, it is more like
    https://github.com/ocornut/imgui/blob/master/examples/example_glfw_opengl3/main.cpp or 
    https://github.com/cimgui/cimgui/blob/docking_inter/backend_test/example_glfw_opengl3/main.c

but for glfw and dx11

**Comment by CybernetHacker14** [comment]
> > Is this the one you are looking for
> 
> No, it is more like https://github.com/ocornut/imgui/blob/master/examples/example_glfw_opengl3/main.cpp or https://github.com/cimgui/cimgui/blob/docking_inter/backend_test/example_glfw_opengl3/main.c
> 
> but for glfw and dx11

Currently, I only have these two files - [ImGuiLayer.cpp](https://github.com/CybernetHacker14/Sentinel/blob/cpp-backup/Engine/Source/Platform/Windows/ImGui/ImGuiLayer.cpp) and [ImGuiBase.cpp](https://github.com/CybernetHacker14/Sentinel/blob/cpp-backup/Scribe/Source/Renderer/ImGuiBase.cpp) at hand, in which the ImGui logic for my editor resided - there isn't a single file like the examples which you have mentioned.

Basically: 
- `ImGuiLayer.cpp` contains the important initialization code required for ImGui - the one time startup and shutdown calls to the `imgui_impl_glfw` and `imgui_impl_dx11` functions and also the per-frame `ImGui::NewFrame()` and `ImGui::Render()` calls.
- `ImGuiBase.cpp` contains the base setup for my editor, where I am drawing the dockspace, the title-bar, and also calling the `ImGuiRender` functions of various other panels which I setup. You can replace all of this with the `ImGui::ShowDemoWindow()` call in order to check for functionality.

So the order of execution of functions is like this - 
1. `ImGuiLayer::OnAttach()`
2. `ImGuiBase::OnAttach()`
-- start of render loop --
3. `ImGuiLayer::Begin()`
4. `ImGuiBase::OnImGuiRender()`
5. `ImGuiLayer::End()
-- end of render loop --
6. `ImGuiBase::OnDetach()`
7. `ImGuiLayer::OnDetach()`

**Comment by sonoro1234** [comment]
I have added main.cpp (dx11 seems to need cpp instead C) and I am getting a compilation without undefined references.
Due to my poor knowledge of dx11 it does not work well but it is a prove of concept about being able to compile and link.

**Comment by CybernetHacker14** [comment]
> I have added main.cpp (dx11 seems to need cpp instead C) and I am getting a compilation without undefined references. Due to my poor knowledge of dx11 it does not work well but it is a prove of concept about being able to compile and link.
> 
> After last commit update it not only compiles but it seems to work correctly

Thanks @sonoro1234 for the implementation. I was able to refactor the existing main.cpp file and have written the C version, you can check and merge the PR #269 if you want regarding this.

**Comment by CybernetHacker14** [comment]
Okay, I also was able to make it work in my project. The reason was hidden very intricately in the `CMakeLists.txt` -> `IMGUI_IMPL_API=extern "C" __declspec(dllexport)` needs to be added to the library defines. I didn't add this define to my premake file. After cloning the branch, and accidentally taking a closer look at the `IMGUI_IMPL_API` macro, I noticed it being different to what I have in my project. Adding it to the premake script is now making it compile and find the definition appropriately.
